### ISSUE TITLE
dotnet run: application exception causes "dotnet has stopped working" error dialog

### ISSUE THREAD

**Issue by TrexinanF14** [dotnet/cli#3993]
OS: Windows
## Steps to reproduce

Create a single line console application that throws an exception. Execute using dotnet run.
## Expected  behavior

Print the exception output && return the correct exit code (as it does currently), but nothing else.
## Actual behavior

Triggers the windows "dotnet has stopped working" error message. This is irksome when I'm in the middle of development and I have to close the error dialog every time my app throws an exception. I believe this can be adjusted in the application code, but it would be much friendlier if each application didn't need to do so.

See https://msdn.microsoft.com/en-us/library/ms680621(VS.85).aspx

**Comment by dhnnjy** [comment]
I encountered similar error today which is forcing Visual studio to restart every time.

dotnet --version
1.0.0-preview2-003131

**Comment by piotrpMSFT** [comment]
The proposal here is a `global exception handler`, generally a bad idea as it prevents crash data from being captured. The current behavior is useful because it allows the process to fail, dumps to get generated, and debug data to get back to developers.  I agree that VS shouldn't crash [please file an issue on https://github.com/aspnet/tooling] but I'm otherwise hard-pressed to see the benefit here.

**Comment by piotrpMSFT** [comment]
An app that wants the behavior described in the issue can simply add a global exception handler in Main to get the same experience.

**Comment by TrexinanF14** [comment]
I just want to clarify a bit to ensure this request is fully captured. 

To quote the dotnet run documentation: "This command is useful for fast iterative development...". I would argue that if this command is intended to help developers quickly iterate on an application, it should avoid behavior that can break that workflow. Popping a windows error dialog every time your application throws an exception definitely goes against this, especially during development when exceptions are common. Requiring _each_ .net developer to add this pinvoke to _each_ application to prevent this behavior is rather painful (especially for fledgling programmers!). 

I could also cite a number of other language cli's that do not display this behavior, if you are looking for a precedent. Python, for example, is one that I consider the gold standard.

I really request you reconsider this from a usability standpoint.

**Comment by piotrpMSFT** [comment]
@TrexinanF14 thanks for the added detail. @blackdwarf can you give this UX issue some thought and reactivate if appropriate?

**Comment by blackdwarf** [comment]
I actually did reconsider this, and I believe that @TrexinanF14 is right: dotnet run should not cause the host to throw on app's exception. This does drive me up the wall at least when coding. Asking me to put in a try-catch block in my Main method is fine, but that is 6 lines of code I have to do *every time* so I don't think it's ideal.

### Plan suggestion
`dotnet run` should catch any and all exceptions from the app and then write out the stack trace to STDERR and then exit with a non-successful exit code (e.g. not 0). This will get the stack trace to the user, the command would signal to the shell that something bad happened but we would not be getting the window on macOS and Windows.

### Caveats
Now, there is a **big** caveat with this plan insofar as the operating system's facilities would not kick in on exception, since `dotnet run` would be handling any and all exceptions from the code it is invoking. This means that there would be no automatic logs in the OS loggers, no crash dumps etc. This would make `dotnet run` usable exclusively on developer machines.

With this, reactivating the issue for further comments and thoughts.

**Comment by TrexinanF14** [comment]
I think this is acceptable behavior. During development, just seeing the error on stderr is usually enough. I don't think it's much to ask for source-distributed apps to run `dotnet build`, then `dotnet <app.dll>`, rather than just `dotnet run`. An added benefit of declaring `dotnet run` a developer-only tool is that any future decisions around this command only needs to worry about developer happiness, rather than worrying about how it will be used in production as well. And it gives all developers a 'right' way to do production deployments (rather than being ambiguous if they should use `dotnet/dotnet build` vs `dotnet run`).

**Comment by svick** [comment]
Doing this would mean that Just-In-Time Debugging with `dotnet run` would no longer be possible, right?

I've found that feature to be useful sometimes, though it doesn't seem to work well currently with .Net Core (you're shown broken inside some framework code, not your app's code).

**Comment by blackdwarf** [comment]
@TrexinanF14 to be fair, `dotnet run` has always been a "dev-only scenario" focused due to the way it worked and how it resolved dependencies. This change would emphasize that further.  I want to see if others see any problem(s) with that. 

@svick how so?

**Comment by svick** [comment]
@blackdwarf As far as I know Just-In-Time Debugging activates when an exception is uncaught by an application. If `dotnet run` caught any exceptions thrown by user code, that would no longer happen and so Just-In-Time Debugging wouldn't activate.

**Comment by blackdwarf** [comment]
@svick ah, that feature, ok. Yes, that would fall into the "any OS (or other higher-level) facilities would not be usable".

**Comment by TrexinanF14** [comment]
@blackdwarf The documentation on the `dotnet run` command explicitly calls out one non-development scenario, so I wasn't sure if it was truly intended to be dev-only even though that was the primary use.
'...and can also be used to run a source-distributed program (for example, a website).'

**Comment by blackdwarf** [comment]
@TrexinanF14 yes, that was my point with the notion that if this change goes in, `dotnet run` would become a pure dev-only concern.
